cmake: add scope support to script mode

pillo79 opened the pull request:
In CMake script mode, "targets" are an undefined concept, therefore no properties can be attached to them; those CMake functions are either silently ignored or result in errors. This means that `zephyr_set()` and any functionality that depends on it (e.g. the `yaml` module) is not usable in script mode.

This PR works around this limitation as #83199 by using `GLOBAL PROPERTY` instead of `TARGET` entities.
Two additional functions are introduced to offer a simplified access to the scope functionality:
* a `zephyr_exists_scope` macro to check if a scope exists; and 
* a `zephyr_get_scoped` function to retrieve scoped property values. 

These utility functions are then used in the current code to replace direct property retrieval and scope checks.

The current test suite for the `zephyr_get` API is modified to also run itself in script mode and detect any future issue.

This, along with #80007, is a prerequisite for the LLEXT EDK refactor in #78727.

pillo79 committed:
tests: cmake: run zephyr_get() tests in script mode

Re-run the zephyr_get() testsuite in script mode after the project
mode testsuite has been executed. This is to ensure that the
zephyr_get() function works correctly in script mode as well.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

tejlmand reviewed (changes requested):
Biggest concern for now is introduction of different implementation for normal vs script mode, as well as `if(CMAKE_SCRIPT_MODE_FILE)` starting to creep into various places in CMake code.

The script mode possibility was introduced for support `package_helper`, see #41302.

Of course purpose can be extended if there is a need, but from the PR description it is not exactly clear to me why we need this change.

Even with this PR, then `yaml_save()` is only called from `<zephyr-base>/CMakeLists.txt` and thus never called from the package helper.

I see some references to other llext related PRs, but even with those it is not fully clear who / why / when the yaml output is required in script mode, as twister itself doesn't seem to be needing it.

Please provide some more description of what is failing and when, and why this is problematic compared to the initial purpose of script mode, so that PR can be reviewed in that context and the extended script mode purpose is clear.

pillo79 reviewed (commented):
> from the PR description it is not exactly clear to me why we need this change.

@tejlmand I feel like I'm on a goose chase :slightly_smiling_face: but here's a recap of the full story: 

* In September I have proposed PR #78727, that limited the rework to the EDK files and the call site in the main `CMakelists.txt`. This was blocked twice, the last time in favor of using the just-introduced YAML support. That lacked a way to expand generator expressions, which is extremely important for the EDK. So I worked on that.

* I have been working since October to add genex support to the YAML module in PR #80007, before reworking the above PR to use it. As you well know, the `yaml` module heavily depends on the scope functionality in `zephyr_set`, and since I now need to use it from `llext-edk.cmake` (which is called in script mode), I need to also add that support. So I worked on that.

* I submitted this PR as the third attempt at moving forward with something, mentioning the relevant threads, and yet this was blocked again because -among other things- there is no obvious need for it. 

I do have a working implementation here, of course. I just pushed an update to the YAML PR yesterday, and I was rewording that PR's description at the time you commented here. Updating production-quality commits and rationales for 3 cascading PRs on every change takes me some time... :cold_sweat:  

I sincerely appreciate your reviews and insightful comments :bow:, and try to apply all of them. In this instance however I really feel it's unfair to dismiss this PR without providing any guidance on how you think the issue should be properly addressed.

---

* Do you see any alternative to adding script mode support for `zephyr_set` et al, to be able to use the `yaml` module in the LLEXT EDK? 
* If we agree on adding this functionality, modifying the existing functions to use `SOURCE` and `TARGET` in the appropriate contexts minimizes code and semantic changes IMO. Do you have any advice on how else to approach this?

Thanks in advance for your time, but please advise how to proceed.

tejlmand commented:
>  I now need to use it from llext-edk.cmake (which is called in script mode), I need to also add that support. 

That was the missing piece of info I needed, for some reason I had missed the combo that llext-edk which is now being updated to use the yaml parser is actually called in script mode. :)

Guess I was to focused on the Zephyr build system itself to remember that the yaml module itself would be re-used.

tejlmand commented:
>  In this instance however I really feel it's unfair to dismiss this PR without providing any guidance on how you think the issue should be properly addressed.

Sorry, I didn't mean any offense in the nack, I think the work done is great, but to provide the best guidance I needed the actual use-case on where the script mode entered the loop.

So it was with a genuine desire to fully understand the flow and not having to review this code bty traversing the other PRs.
I could see they were connected, but not the exact connection point (llext-edk.cmake)

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -3359,7 +3359,12 @@ function(zephyr_get_scoped output scope variable)
   endif()
 
-  get_property(value TARGET ${scope}_scope PROPERTY ${variable})
+  if(CMAKE_SCRIPT_MODE_FILE)
```
> if this was a collection of separate files it would be easier to cherry-pick and provide alternatives, but that is yet another sizable effort

another one on my wish list, but before moving all those functions to multiple other locations / individual CMake modules, then we must have proper generated docs, else knowing where to find a given function will become even harder.

tejlmand committed:
cmake: use GLOBAL property instead TARGET properties for scoping

Targets are not available in script mode.
To support the Zephyr scoping feature used by snippets and yaml module
then this commit moves from using custom targets to use GLOBAL
properties for scopes.

A scope property is prefixed with `<scope>:<property>` to avoid naming
collisions.
A `scope:<scope-name>` global property is used to track created scopes.
Tracking valid scopes ensure that properties are only set on known
scopes and thus catches typos / naming errors.

Add zephyr_scope_exists() and zephyr_get_scoped() to abstract the
implementation details of the scoped property retrieval and refactor
current code to use them.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>
Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

tejlmand commented:
@pillo79 made a quick draft here https://github.com/zephyrproject-rtos/zephyr/pull/83199 where I switched from using targets and target properties to instead using global properties.

To still support scope creation / checking, which I think is quite important for the ability to catch mistake like scope name typos / naming refactory, then I've kept the feature of tracking created scopes.

Take a look, I think the principle used there can be merged into this work.
And a big bonus is that we avoid script mode checking as the scoping feature now works in both normal and script mode.

Feel free to take and modify https://github.com/zephyrproject-rtos/zephyr/pull/83199 anyway you like so that it suits your needs.

pillo79 commented:
Thanks @tejlmand, I did not dare touch how `zephyr_set` worked :grin: 

I amended your commit with the `zephyr_get_scoped` functionality. 
I also left the `build_info` script mode stub because the `yaml` module is not loaded by `extensions.cmake` (should it?)
Ran the `cmake.zephyr_get` test suite again so this should be good now!

tejlmand commented:
>  I also left the build_info script mode stub because the yaml module is not loaded by extensions.cmake (should it?)

Good point. I tested my work locally with package helper and some extra local mods for build_info and discovered that we could avoid the stub in those cases :smile: 

But you're right, we should not load `yaml` module in extensions.cmake. I would defer that until next step cleanup https://github.com/zephyrproject-rtos/zephyr/pull/83045#discussion_r1890063712

pillo79 commented:
> this should be good now!

^-- only slightly less effective than "what can possibly go wrong"
:bomb: 

Fixed an issue with `APPEND` property to `zephyr_set`.

tejlmand reviewed (approved):
Nice catch on the argument order swap :see_no_evil:
